client-server-server/src/main/scala/zio/raft/server/RaftServer.scala

```diff
@@ -0,0 +1,473 @@
+   */
+  case class Sessions(
+    metadata: Map[SessionId, SessionMetadata],
+    connections: Map[SessionId, SessionConnection],
+    routingToSession: Map[RoutingId, SessionId]
+  ) {
+    def addConnected(sessionId: SessionId, routingId: RoutingId, capabilities: Map[String, String], now: Instant): Sessions = {
+      val expiresAt = now.plusSeconds(90) // TODO: use config
+      copy(
+        metadata = metadata.updated(sessionId, SessionMetadata(capabilities, now)),
+        connections = connections.updated(sessionId, SessionConnection(Some(routingId), expiresAt)),
+        routingToSession = routingToSession.updated(routingId, sessionId)
+      )
+    }
+    
+    def reconnect(sessionId: SessionId, routingId: RoutingId, now: Instant): Sessions = {
+      connections.get(sessionId) match {
+        case Some(conn) =>
+          val expiresAt = now.plusSeconds(90) // TODO: use config
+          // Remove old routing mapping if it exists
+          val cleanedRouting = conn.routingId.map(old => routingToSession.removed(old)).getOrElse(routingToSession)
+          copy(
+            connections = connections.updated(sessionId, conn.copy(routingId = Some(routingId), expiresAt = expiresAt)),
+            routingToSession = cleanedRouting.updated(routingId, sessionId)
+          )
+        case None => this
+      }
+    }
+    
+    def disconnect(sessionId: SessionId, routingId: RoutingId): Sessions =
+      copy(
+        connections = connections.updatedWith(sessionId)(_.map(_.copy(routingId = None))),
+        routingToSession = routingToSession.removed(routingId)
+      )
+    
+    def updateExpiry(sessionId: SessionId, now: Instant): Sessions = {
+      connections.get(sessionId) match {
+        case Some(conn) =>
+          val expiresAt = now.plusSeconds(90) // TODO: use config
+          copy(connections = connections.updated(sessionId, conn.copy(expiresAt = expiresAt)))
+        case None => this
+      }
+    }
+    
+    def removeExpired(now: Instant): (List[SessionId], Sessions) = {
+      val expired = connections.filter { case (_, conn) => conn.expiresAt.isBefore(now) }.keys.toList
+      val newMetadata = metadata -- expired
+      val newConnections = connections -- expired
+      val newRouting = routingToSession.filter { case (_, sid) => !expired.contains(sid) }
+      (expired, Sessions(newMetadata, newConnections, newRouting))
+    }
+    
+    def getRoutingId(sessionId: SessionId): Option[RoutingId] =
+      connections.get(sessionId).flatMap(_.routingId)
+    
+    def getMetadata(sessionId: SessionId): Option[SessionMetadata] =
+      metadata.get(sessionId)
+    
+    def findSessionByRouting(routingId: RoutingId): Option[SessionId] =
+      routingToSession.get(routingId)
+    
+    def closeAll(transport: ZmqServerTransport, leaderId: Option[MemberId]): UIO[Unit] = {
+      val reason = if (leaderId.isDefined) SessionCloseReason.NotLeaderAnymore else SessionCloseReason.Shutdown
```

Comment:
not good enough, we can be not a leader anymore even when we don't know the leader

Reply:
we should have two variants of this method: shutdown and stepdown